src/sinks/syslog.rs

```diff
@@ -0,0 +1,594 @@
+            Mode::Udp(config) => config.build(cx, move |e| Some(syslog_encode(e, false))),
+            #[cfg(unix)]
+            Mode::Unix(config) => config.build(cx, move |e| Some(syslog_encode(e, true))),
+        }
+    }
+
+    fn input_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn sink_type(&self) -> &'static str {
+        "syslog"
+    }
+}
+
+fn build_syslog_message(
+    event: Event,
+    include_len: bool,
+    format: Format,
+    include_extra_fields: bool,
+    appname_key: &str,
+    facility_key: &str,
+    host_key: &str,
+    msgid_key: &str,
+    procid_key: &str,
+    severity_key: &str,
+    default_facility: SyslogFacility,
+    default_severity: SyslogSeverity,
+) -> Bytes {
+    let mut log = event.into_log();
+
+    let ts = log.remove(log_schema().timestamp_key()).and_then(|v| {
+        v.as_timestamp()
+            .map(|v| FixedOffset::west(0).timestamp_nanos(v.timestamp_nanos()))
+    });
+
+    let procid = log
+        .remove(procid_key)
+        .map(|procid| match procid {
+            Value::Integer(pid) => ProcId::PID(pid as i32),
+            Value::Bytes(_) => ProcId::Name(procid.to_string_lossy()),
+            _ => ProcId::Name("vector".to_string()),
+        })
+        .or_else(|| Some(ProcId::Name("vector".to_string())));
+
+    let msg = Message {
+        procid,
+        protocol: format.into(),
+        timestamp: ts,
+        facility: log
+            .remove(facility_key)
+            .and_then(|v| v.into())
+            .or(Some(default_facility)),
+        severity: log
+            .remove(severity_key)
+            .and_then(|v| v.into())
+            .or(Some(default_severity)),
+        appname: log.remove(appname_key).map(|v| v.to_string_lossy()),
+        msgid: log.remove(msgid_key).map(|v| v.to_string_lossy()),
+        hostname: log.remove(host_key).map(|v| v.to_string_lossy()),
+        msg: log
+            .remove(log_schema().message_key())
+            .map(|v| v.to_string_lossy())
+            .unwrap_or_else(|| "-".to_owned()),
+        structured_data: if include_extra_fields {
+            build_structured_data(log)
+        } else {
+            vec![]
+        },
+    };
+
+    if include_len {
+        let msg = format!("{}", msg);
+        Bytes::from(format!("{} {}\n", msg.len(), msg))
+    } else {
+        Bytes::from(format!("{}\n", msg))
+    }
+}
+
+fn build_structured_data(log: LogEvent) -> Vec<StructuredElement<String>> {
+    let mut d = vec![];
+    for (k, v) in log.as_map().iter() {
+        let mut e = StructuredElement {
+            id: k.clone(),
+            params: vec![],
+        };
+        if let Value::Map(m) = v {
+            for (k, v) in m.iter() {
+                e.params.push((k.clone(), v.to_string_lossy()));
+            }
+        } else {
+            e.params.push(("value".to_string(), v.to_string_lossy()));
+        }
+        d.push(e);
+    }
+    d
```

Comment:
This looks like it could be better structured using maps:
```suggestion
    log.as_map().iter().map(|(k, v)| {
        StructuredElement {
            id: k.clone(),
            params: match v {
                Value::Map(m) => {
                    m.iter().map(|(k, v)| (k.clone(), v.to_string_lossy())).collect()
                }
                v => vec![("value".to_string(), v.to_string_lossy())],
            },
        }
    })
    .collect()
```